executor/analyze.go

```diff
@@ -496,22 +527,469 @@ type AnalyzeFastTask struct {
+			e.sampCursor++
+		}
+
+		err = e.updateCollectorSamples(iter.Value(), iter.Key(), p, hasPKInfo)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func (e *AnalyzeFastExec) handleScanTasks(bo *tikv.Backoffer) error {
+	snapshot, err := e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	if err != nil {
+		return err
+	}
+	for _, t := range e.scanTasks {
+		iter, err := snapshot.Iter(t.StartKey, t.EndKey)
+		if err != nil {
+			return err
+		}
+		err = e.handleScanIter(iter)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleSampTasks(bo *tikv.Backoffer, rid int, err *error) {
+	defer func() {
+		e.wg.Done()
+	}()
+	var snapshot kv.Snapshot
+	snapshot, *err = e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	rander := rand.New(rand.NewSource(RandSeed + int64(rid)))
+	if *err != nil {
+		return
+	}
+	for i := rid; i < len(e.sampTasks); i += e.concurrency {
+		task := e.sampTasks[i]
+		if task.SampSize == 0 {
+			continue
+		}
+
+		var tableID, minRowID, maxRowID int64
+		startKey, endKey := task.Location.StartKey, task.Location.EndKey
+		tableID, minRowID, *err = tablecodec.DecodeRecordKey(startKey)
+		if *err != nil {
+			return
+		}
+		_, maxRowID, *err = tablecodec.DecodeRecordKey(endKey)
+		if *err != nil {
+			return
+		}
+		if maxRowID <= minRowID {
+			continue
+		}
+
+		keys := make([]kv.Key, 0, task.SampSize)
+		for i := 0; i < int(task.SampSize); i++ {
+			randKey := rander.Int63n(maxRowID-minRowID) + minRowID
+			keys = append(keys, tablecodec.EncodeRowKeyWithHandle(tableID, randKey))
+		}
+
+		var kvMap map[string][]byte
+		for _, key := range keys {
+			var iter kv.Iterator
+			iter, *err = snapshot.Iter(key, endKey)
+			if *err != nil {
+				return
+			}
+			kvMap[string(iter.Key())] = iter.Value()
+		}
+
+		*err = e.handleBatchGetResponse(kvMap)
+		if *err != nil {
+			return
+		}
+	}
+}
+
+func (e *AnalyzeFastExec) buildHist(ID int64, collector *statistics.SampleCollector, tp *types.FieldType) (*statistics.Histogram, error) {
+	// TODO: build histogram and cmsketch here for one collector.
+	return nil, nil
+}
+
+func (e *AnalyzeFastExec) runTasks() ([]*statistics.Histogram, []*statistics.CMSketch, error) {
+	errs := make([]error, e.concurrency)
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	// collect column samples and primary key samples and index samples.
+	length := len(e.colsInfo) + hasPKInfo + len(e.idxsInfo)
+	e.collectors = make([]*statistics.SampleCollector, length)
+	for i := range e.collectors {
+		e.collectors[i] = &statistics.SampleCollector{
+			FMSketch:      statistics.NewFMSketch(maxSketchSize),
+			MaxSampleSize: int64(MaxSampleSize),
+			CMSketch:      statistics.NewCMSketch(defaultCMSketchDepth, defaultCMSketchWidth),
+			Samples:       make([]*statistics.SampleItem, MaxSampleSize),
+		}
+	}
+
+	e.wg.Add(e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		bo := tikv.NewBackoffer(context.Background(), 500)
```

Comment:
can we only build one Backoffer and use it in many goroutines? like what we did in `buildSampTask()`